File: RIS/procedural/spiralStaircaseCpp/spiral.cpp
#include <iostream>
#include <string>
#include <regex>
#include <iterator>
#include <algorithm>
#include <fstream>
#include <ios>
/*
# Signal to prman that we have finished (\377) and
# print out error message on stderr
*/
void niceExit(const std::string &_message) 
{
  std::cerr<<_message;
  //flush stream so prman knows we are done
  std::cout<<'\377';
  std::cout<<std::flush; 
  exit(EXIT_FAILURE);
}	

void block(float _w, float _h, float _d)
{
  std::cout<<"TransformBegin\n";
	float w=_w/2.0f;
	float h=_h/2.0f;
	float d=_d/2.0f;
	//#rear face
	std::cout<<"Translate "<<w<<' '<<h<<" 0\n";
	std::cout<<"\tPatch 'bilinear' 'P' [" << -w<<' '<< -h<<' '<<d<<' '
                                      << -w<<' '<<  h<<' '<<d<<' '
                                      <<  w<<' '<< -h<<' '<<d<<' '
                                      <<  w<<' '<<  h<<' '<<d<<"]\n";
  																	

								
//	#front face
	std::cout<<"\tPatch 'bilinear' 'P' [" << -w<<' '<< -h<<' '<<-d<<' '
                                      << -w<<' '<<  h<<' '<<-d<<' '
                                      <<  w<<' '<< -h<<' '<<-d<<' '
                                      <<  w<<' '<<  h<<' '<<-d<<"]\n";
  
	// #left face
	std::cout<<"\tPatch 'bilinear' 'P' [" << -w<<' '<< -h<<' '<<-d<<' '
                                      << -w<<' '<<  h<<' '<<-d<<' '
                                      << -w<<' '<< -h<<' '<< d<<' '
                                      <<  -w<<' '<< h<<' '<< d<<"]\n";
  
	// #right face
	std::cout<<"\tPatch 'bilinear' 'P' [" <<  w<<' '<< -h<<' '<<-d<<' '
                                      <<  w<<' '<<  h<<' '<<-d<<' '
                                      <<  w<<' '<< -h<<' '<< d<<' '
                                      <<  w<<' '<<  h<<' '<< d<<"]\n";
  
 								
	// #bottom face
	std::cout<<"\tPatch 'bilinear' 'P' [" <<  w<<' '<< -h<<' '<< d<<' '
                                      <<  w<<' '<< -h<<' '<<-d<<' '
                                      << -w<<' '<< -h<<' '<< d<<' '
                                      << -w<<' '<< -h<<' '<<-d<<"]\n";

								
	
	// #top face
	std::cout<<"\tPatch 'bilinear' 'P' [" <<  w<<' '<<  h<<' '<< d<<' '
                                      <<  w<<' '<<  h<<' '<<-d<<' '
                                      << -w<<' '<<  h<<' '<< d<<' '
                                      << -w<<' '<<  h<<' '<<-d<<"]\n";

									
	std::cout<<"TransformEnd\n";

}


int main()
{
  // string to store the arguments sent from prman
  std::string args;
  // regex for removing whitespace from input.
  std::regex ws_re("\\s+"); // whitespace
  // so we need to loop until no input on cin.
  while(  std::getline(std::cin, args))
  {
    // this will store the input tokens for conversion later
    // if we need to do arithmetic on them they will need conversion
    std::vector<std::string> tokens;

    // copy the tokens into a vector
    std::copy( 
              std::sregex_token_iterator(args.begin(), 
              args.end(), ws_re, -1),
              std::sregex_token_iterator(),
              std::back_inserter(tokens)
              );
    // ensure we got enough tokens for this demo
    if(tokens.size() <6)
    {
      niceExit("Error in ribfile not enough arguments");
    }
    float width=std::stof(tokens[1]);
    float height=std::stof(tokens[2]);
    float depth=std::stof(tokens[3]);
    float stairHeight=std::stof(tokens[4]);
    float rotationAngle=std::stof(tokens[5]);

    // Now we will output the cylinder for the center.
    //  # do an explicit TransformBegin
    std::cout<<"TransformBegin\n";
    //			# now the center cylinder
    std::cout<<"\tTranslate 0 -1 0\n";
    std::cout<<"\tRotate -90 1 0 0\n";
    //			size=StairHeight+(3*height)
    std::cout<<"\tCylinder 0.2 "<< stairHeight<< " 1 360\n" ;
    std::cout<<"TransformEnd\n";


  // # now we do some calculations and make our spiral
    float i=0.0f;
    std::cout<<"TransformBegin\n";
    // # loop and build the blocks
    while (i <stairHeight)
    {
      block(width,height,depth);
      std::cout<<"\tTranslate 0 "<<height<<" 0\n";
      std::cout<<"\tRotate "<<rotationAngle<<" 0 1 0\n";
      i+=height;
    }
    std::cout<<"TransformEnd\n";

  //flush stream so prman knows we are done
    std::cout<<'\377';
    std::cout<<std::flush; 
    args.clear();
  }
}

File: RIS/procedural/spiralStaircasePython/spiral.py
#!/usr/bin/python

######################################################################
# spiral.py Renderman Procedural Usage
# Procedural 'RunProgram' ['spiral.py' 'w h d H A'] [Bounding Box]
# Parameters Width Height Depth of Stair block
# Height of stairs
# Angle of stair Rotation
#
######################################################################


import sys

# grab the function ptr to write should speed things up.
Write = sys.stdout.write

######################################################################
# Function to draw a block made from Patches
# The center of the block if w/2 d/2 h/2
######################################################################


def Block(width, hight, depth):
    Write("TransformBegin\n")
    w = width / 2.0
    h = hight / 2.0
    d = depth / 2.0
    # rear face
    Write("Translate %f %f 0" % (w, h))
    Write("	Patch 'bilinear' 'P' [%f %f %f %f %f %f %f %f %f %f %f %f   ]" % (-w, -h, d, -w, h, d, w, -h, d, w, h, d))
    # front face
    Write(
        "	Patch 'bilinear' 'P' [%f %f %f %f %f %f %f %f %f %f %f %f   ]" % (-w, -h, -d, -w, h, -d, w, -h, -d, w, h, -d)
    )

    # left face
    Write(
        "	Patch 'bilinear' 'P' [%f %f %f %f %f %f %f %f %f %f %f %f   ]" % (-w, -h, -d, -w, h, -d, -w, -h, d, -w, h, d)
    )
    # right face
    Write("	Patch 'bilinear' 'P' [%f %f %f %f %f %f %f %f %f %f %f %f   ]" % (w, -h, -d, w, h, -d, w, -h, d, w, h, d))
    # bottom face
    Write(
        "	Patch 'bilinear' 'P' [%f %f %f %f %f %f %f %f %f %f %f %f   ]" % (w, -h, d, w, -h, -d, -w, -h, d, -w, -h, -d)
    )

    # top face
    Write("	Patch 'bilinear' 'P' [%f %f %f %f %f %f %f %f %f %f %f %f   ]" % (w, h, d, w, h, -d, -w, h, d, -w, h, -d))
    Write("TransformEnd\n")


###############################################
# Signal to prman that we have finished (\377) and
# print out error message on stderr
################################################
def NiceExit(message):
    sys.stderr.write(message)
    # flush stream so prman knows we are done
    Write("\377")
    sys.stdout.flush()


###############################################################################################
# prman RunPrograms must read in a the data from stdin in the format
# The detail is provided first, as a single floating-point number, followed by a space,
# followed by the datablock and finally a newline.
# The datablock is completely uninterpreted by the renderer or by the socket write,
# so any newlines or special characters should be preserved
# (but the quotation marks that make it a string in the RIB file will, of course, be missing).
###############################################################################################

if __name__ == "__main__":
    # get the datablock from prman
    args = sys.stdin.readline()
    # now loop round as we may get more data
    while args:
        # split the  data into blocks
        words = args.split()
        # the first is the detail (a float)
        detail = float(words[0])
        # now check  to see if we have enough arguments
        if len(words) < 6:
            NiceExit("Error in ribfile not enough arguments\n Args were %s" % args)
            break
        # so we have enough arguments so extract the detail we need
        else:
            # first get the datablock from prman
            datablock = words[1:]
            # and extract the data we need
            width = float(datablock[0])
            height = float(datablock[1])
            depth = float(datablock[2])
            StairHeight = float(datablock[3])
            RotationAngle = float(datablock[4])
            # do an explicit TransformBegin
            Write("TransformBegin\n")
            # now the center cylinder
            Write("Translate 0 -1 0\n")
            Write("Rotate -90 1 0 0\n")
            size = StairHeight + (3 * height)
            Write("Cylinder 0.2  %f 1 360\n" % (size))
            sys.stderr.write("%f \n" % (size))
            Write("TransformEnd\n")

            # now we do some calculations and make our spiral
            i = 0.0
            Write("TransformBegin\n")
            # loop and build the blocks
            while i <= StairHeight:
                Block(width, height, depth)
                Write("Translate 0 %f 0\n" % height)
                Write("Rotate %f 0 1 0\n" % RotationAngle)
                i += height
            Write("TransformEnd\n")

            # we now close the rib stream and get the  next datablock  if there  is  one

            Write("\377")
            sys.stdout.flush()
        # read the next line
        args = sys.stdin.readline()
